vpr/src/route/binary_heap.cpp

```diff
@@ -0,0 +1,184 @@
+#include "binary_heap.h"
+#include "vtr_log.h"
+
+static size_t parent(size_t i) { return i >> 1; }
+// child indices of a heap
+static size_t left(size_t i) { return i << 1; }
+static size_t right(size_t i) { return (i << 1) + 1; }
+
+BinaryHeap::BinaryHeap()
+    : heap_(nullptr)
+    , heap_size_(0)
+    , heap_tail_(0) {}
+
+BinaryHeap::~BinaryHeap() {
+    free_all_memory();
+}
+
+t_heap* BinaryHeap::alloc() {
+    return storage_.alloc();
+}
+void BinaryHeap::free(t_heap* hptr) {
+    storage_.free(hptr);
+}
+
+// Note: malloc()/free() must be used for the heap,
+//       or realloc() must be eliminated from add_to_heap()
+//       because there is no C++ equivalent.
+void BinaryHeap::init_heap(const DeviceGrid& grid) {
+    ssize_t target_heap_size = (grid.width() - 1) * (grid.height() - 1);
+    if (heap_ == nullptr || heap_size_ < target_heap_size) {
+        if (heap_ != nullptr) {
+            vtr::free(heap_ + 1);
+            heap_ = nullptr;
+        }
+        heap_size_ = (grid.width() - 1) * (grid.height() - 1);
+        heap_ = (t_heap**)vtr::malloc(heap_size_ * sizeof(t_heap*));
+        heap_--; /* heap stores from [1..heap_size] */
+    }
+    heap_tail_ = 1;
+}
+
+void BinaryHeap::add_to_heap(t_heap* hptr) {
+    expand_heap_if_full();
+    // start with undefined hole
+    ++heap_tail_;
+    sift_up(heap_tail_ - 1, hptr);
+}
+
+bool BinaryHeap::is_empty_heap() const {
+    return (bool)(heap_tail_ == 1);
+}
+
+t_heap* BinaryHeap::get_heap_head() {
+    /* Returns a pointer to the smallest element on the heap, or NULL if the     *
+     * heap is empty.  Invalid (index == OPEN) entries on the heap are never     *
+     * returned -- they are just skipped over.                                   */
+
+    t_heap* cheapest;
+    size_t hole, child;
+
+    do {
+        if (heap_tail_ == 1) { /* Empty heap. */
+            VTR_LOG_WARN("Empty heap occurred in get_heap_head.\n");
+            return (nullptr);
+        }
+
+        cheapest = heap_[1];
+
+        hole = 1;
+        child = 2;
+        --heap_tail_;
+        while ((int)child < heap_tail_) {
+            if (heap_[child + 1]->cost < heap_[child]->cost)
+                ++child; // become right child
+            heap_[hole] = heap_[child];
+            hole = child;
+            child = left(child);
+        }
+        sift_up(hole, heap_[heap_tail_]);
+
+    } while (cheapest->index == OPEN); /* Get another one if invalid entry. */
+
+    return (cheapest);
+}
+
+void BinaryHeap::empty_heap() {
+    for (int i = 1; i < heap_tail_; i++)
+        free(heap_[i]);
+
+    heap_tail_ = 1;
+}
+
+size_t BinaryHeap::size() const { return static_cast<size_t>(heap_tail_ - 1); } // heap[0] is not valid element
+
+// make a heap rooted at index i by **sifting down** in O(lgn) time
```

Comment:
Not sure what index i is here.  Do you mean hole is the index where the heap will be rooted?

Reply:
This comment is from the old code, I just copied it here.

Reply:
I think the fix is simply replacing `index i` with `index hole`.